collectors/ApicFaultsCollector.py

```diff
@@ -0,0 +1,169 @@
+        g_apic_faults_type.add_metric(labels=['management'],
+                                value=apic_faults_type_management)
+        g_apic_faults_type.add_metric(labels=['network'],
+                                value=apic_faults_type_network)
+        g_apic_faults_type.add_metric(labels=['operational'],
+                                value=apic_faults_type_operational)
+        """faults by domain"""
+        apic_faults_domain_infra = 0
+        apic_faults_domain_tenant = 0
+        apic_faults_domain_access = 0
+        apic_faults_domain_external = 0
+        apic_faults_domain_framework = 0
+        apic_faults_domain_security = 0
+        apic_faults_domain_management = 0
+        apic_faults_domain_plugin = 0
+        for fault_object in data['imdata']:
+            if fault_object['faultInst']['attributes']['domain'] == 'infra':
+                apic_faults_domain_infra =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'tenant':
+                apic_faults_domain_tenant =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'access':
+                apic_faults_domain_access =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'external':
+                apic_faults_domain_external =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'framework':
+                apic_faults_domain_framework =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'security':
+                apic_faults_domain_security =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'management':
+                apic_faults_domain_management =+ 1
+            elif fault_object['faultInst']['attributes']['domain'] == 'plugin':
+                apic_faults_domain_plugin =+ 1
+        g_apic_faults_domain.add_metric(labels=['infra'],
+                                value=apic_faults_domain_infra)
+        g_apic_faults_domain.add_metric(labels=['tenant'],
+                                value=apic_faults_domain_tenant)
+        g_apic_faults_domain.add_metric(labels=['access'],
+                                value=apic_faults_domain_access)
+        g_apic_faults_domain.add_metric(labels=['external'],
+                                value=apic_faults_domain_external)
+        g_apic_faults_domain.add_metric(labels=['framework'],
+                                value=apic_faults_domain_framework)
+        g_apic_faults_domain.add_metric(labels=['security'],
+                                value=apic_faults_domain_security)
+        g_apic_faults_domain.add_metric(labels=['management'],
+                                value=apic_faults_domain_management)
+        g_apic_faults_domain.add_metric(labels=['plugin'],
+                                value=apic_faults_domain_plugin)
+        return [g_apic_faults_severity,g_apic_faults_type,g_apic_faults_domain,g_apic_faults_total,g_apic_faults_ack]
```

Comment:
The current four metrics will result in some loss of detail e.g. it won't be possible to filter for `domain` _and_ `severity`.
Also all datapoints are looped over 4 times to create the prometheus metrics.
I would suggest to create one metric that has labels for `severity`, `type`, `domain` and `ack`. This will bring more flexibility when querying later on.

The suggestion will only generate metrics if a combination of `severity`, `type`, `domain` and `ack` was present. So no faults == no metrics. If that is not desired the `add_metric` must be executed for every combination of 'severity`, `type`, `domain` and `ack` which will lead to 6 * 8 * 8 * 2 = 768 time series though.  


```suggestion
    def get_metrics(self, host: str, data: Dict) -> List[GaugeMetricFamily]:
        """Collect and sort APIC Faults by type, severity, domain and acknowledgement"""
        g_apic_faults = GaugeMetricFamily('network_apic_faults_severity',
                                          'APIC faults by severity, type and domain',
                                          labels=['severity', 'type', 'domain', 'ack'])
        g_apic_faults_total = GaugeMetricFamily('network_apic_faults_total',
                                                'APIC faults total')

        """faults total"""
        apic_faults_total = data['totalCount']
        g_apic_faults_total.add_metric(labels=['total'],
                                       value=apic_faults_total)

        class FaultKey:
            self.severity = ''
            self.type = ''
            self.domain = ''
            self.ack = ''

            def __init__(self, severity: str, type: str, domain: str, ack: str) -> None:
                self.severity = severity
                self.type = type
                self.domain = domain
                self.ack = ack

            def __hash__(self):
                return hash((self.severity, self.type, self.domain, self.ack))

            def __eq__(self, other):
                return (self.severity, self.type, self.domain, self.ack) == (other.severity, other.type, other.domain, self.ack)

            def __ne__(self, other):
                return not (self == other)

            def labels(self) -> List[str]:
                return [self.severity, self.type, self.domain, self.ack]

        faults = {}
        for fault_object in data['imdata']:

            severity = fault_object['faultInst']['attributes']['severity']
            faultType = fault_object['faultInst']['attributes']['type']
            domain = fault_object['faultInst']['attributes']['domain']
            ack = fault_object['faultInst']['attributes']['ack']
            key = FaultKey(severity, faultType, domain, ack)
            faults[key] = faults.get(key, 0) + 1
        for k, v in faults.items():
            g_apic_faults.add_metric(labels=k.labels(),
                                     value=v)

        return [g_apic_faults]
```